### PR TITLE
teach inductor to handle floor

### PR DESCRIPTION
**Pull request by ngimel**
Per title, happen when there's upsampling with non-integer scale. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -72,6 +72,27 @@ def _print_CleanDiv(self, expr):
...
+        mod = self.paren(self.doprint(mod))
+        if div != "1":
+            x = f"({x} // {div})"
+        return f"{x} % {mod}"
+
+    def _print_FloorDiv(self, expr):
+        x, div = expr.args
+        x = self.paren(self.doprint(x))
+        div = self.paren(self.doprint(div))
+        return f"({x} // {div})"
```
Comment (ezyang) [inline torch/_inductor/codegen/common.py:89]: Is the outer parentheses here necessary?

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -74,24 +74,14 @@ def is_aligned(x):
...
-
-    def _print_FloorDiv(self, expr):
-        x, div = expr.args
-        x = self.paren(self.doprint(x))
-        div = self.paren(self.doprint(div))
-        return f"({x} // {div})"
+class TritonPrinter(PythonPrinter):
+    def _print_floor(self, expr):
+        assert len(expr.args) == 1
+        return f"tl.libdevice.floor({self.paren(self._print(expr.args[0]))})"
```
Comment (ezyang) [inline torch/_inductor/codegen/triton.py:80]: wanna do ceiling too?

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -1228,10 +1218,10 @@ def call_kernel(self, code, name: str):
         # TODO(jansel): if there are constants, we shouldn't bother passing them as args
         for tree in self.range_trees:
             if isinstance(tree.numel, (sympy.Integer, sympy.Symbol)):
-                expr = texpr(tree.numel)
+                expr = pexpr(tree.numel)
             else:
                 expr = f"{name}_{tree.prefix}numel"
-                code.writeline(f"{expr} = {texpr(tree.numel)}")
+                code.writeline(f"{expr} = {pexpr(tree.numel)}")
```
Comment (ezyang) [inline torch/_inductor/codegen/triton.py:1224]: I didn't audit spots where we currently call texpr and should call pexpr, trusting you to have gotten all of these

File: torch/nn/functional.py [hunk 4]
```diff
@@ -3917,7 +3917,7 @@ def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optiona
                            for i in range(dim)]
         else:
             output_size = [
-                _sym_int(math.floor(_sym_float(input.size(i + 2)) * scale_factors[i]))
+                _sym_int((input.size(i + 2) * scale_factors[i]))
```
Comment (ezyang) [inline torch/nn/functional.py:3920]: nit: redundant parens

**Review by ezyang — APPROVED** [review]
a test for the torch.nn.functional change would be appreciated

**Comment by ngimel** [comment]
nn.functional change should be tested by existing tests (like, they should start failing if I goofed), it's meant to be semantics-preserving, I'm just removing extra wraps that would be redone when calling sym_int or `*` anyway

File: torch/nn/functional.py (hunk 4 above)
Comment (ngimel) [inline torch/nn/functional.py:3920]:
```suggestion
                _sym_int(input.size(i + 2) * scale_factors[i])
```

File: torch/_inductor/codegen/common.py (hunk 1 above)
Comment (ngimel) [reply torch/_inductor/codegen/common.py:89]: Probably not, but it's unchanged from existing.

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (ngimel) [reply torch/_inductor/codegen/triton.py:80]: I was going to look at the cases of ceiling first and see if they need to be added to a printer, will likely end up adding it too. This is a big hammer that will cover our simplification failures, so I'll look into simplifications for a little bit more.

File: torch/_inductor/codegen/triton.py (hunk 3 above)
Comment (ngimel) [reply torch/_inductor/codegen/triton.py:1224]: In other places we import `pexpr` from `codegen.wrapper`, it has been updated.

**Comment by ezyang** [comment]
Re test, I guess I was expecting an inductor opinfo test for upsample to start passing.

**Comment by ngimel** [comment]
Good point, they've started passing even w/o this PR, but since they are not xfailed I didn't notice. 
CPU tests fail with large mismatches, cc @jgong5

File: torch/nn/functional.py (hunk 4 above)
Comment (ngimel) [inline torch/nn/functional.py:3920]:
```suggestion
                _sym_int(input.size(i + 2) * scale_factors[i])
```

**Comment by tiandiao123** [comment]
is it fixed?

**Pull request closed without merging**